src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,328 @@
+
+    state.max_simultaneous_allocations = options.max_simultaneous_allocations;
+
+    const size_t page_size = getPlatformPageSize();
+    state.page_size = page_size;
+    state.slot_size = page_size * options.slot_size;
+
+    /// check that page size is a power of 2
+    assert(isPowerOfTwo(page_size));
+    size_t pool_size
+        = page_size * (1 + state.max_simultaneous_allocations) + state.max_simultaneous_allocations * state.maximumAllocationSize();
+
+    assert(pool_size % page_size == 0);
+    void * guarded_pool_memory = reserveGuardedPool(pool_size);
+    state.guarded_page_pool = reinterpret_cast<uintptr_t>(guarded_pool_memory);
+    state.guarded_page_pool_end = reinterpret_cast<uintptr_t>(guarded_pool_memory) + pool_size;
+
+    size_t free_slots_required_bytes = roundUpTo(state.max_simultaneous_allocations * sizeof(*free_slots), page_size);
+    assert((free_slots_required_bytes % state.page_size) == 0);
+    free_slots = reinterpret_cast<size_t *>(map(free_slots_required_bytes));
+
+    size_t metadata_required_bytes = roundUpTo(state.max_simultaneous_allocations * sizeof(*metadata), page_size);
+    assert((metadata_required_bytes % state.page_size) == 0);
+    metadata = reinterpret_cast<AllocationMetadata *>(map(metadata_required_bytes));
+
+    /// Multiply the sample rate by 2 to give a good, fast approximation for (1 /
+    /// sample_rate) chance of sampling.
+    if (options.sample_rate != 1)
+        adjusted_sample_rate_plus_one = static_cast<uint32_t>(options.sample_rate) * 2 + 1;
+    else
+        adjusted_sample_rate_plus_one = 2;
+
+    getThreadLocals()->next_sample_counter = (getRandomUint32T() % (adjusted_sample_rate_plus_one - 1)) + 1;
+    allocator_ptr = this;
+}
+
+void GuardedPoolAllocator::lock()
+{
+    getThreadLocals()->recursive_guard = true;
+    /// Lock mutex to lock any in-progress allocations
+    pool_reservation_mutex.lock();
+}
+
+void * GuardedPoolAllocator::allocate(size_t size, size_t alignment)
+{
+    if (size > state.maximumAllocationSize())
+        return nullptr;
+
+    if (state.guarded_page_pool_end == 0)
+    {
+        getThreadLocals()->next_sample_counter = adjusted_sample_rate_plus_one - 1;
+        return nullptr;
+    }
+
+
+    /// Protect against recursivity.
+    if (getThreadLocals()->recursive_guard)
+        return nullptr;
+    ScopedRecursiveGuard recursive_guard;
+
+    if (alignment == 0)
+        alignment = alignof(max_align_t);
+
+    size_t slot_index;
+    {
+        std::lock_guard guard{pool_reservation_mutex};
+        slot_index = reserveSlotInGuardedPool();
+    }
+
+    if (slot_index == kInvalidSlotID)
+    {
+        return nullptr;
+    }
+
+    uintptr_t slot_start = state.slotToAddr(slot_index);
+    uintptr_t slot_end = state.slotToAddr(slot_index) + state.maximumAllocationSize();
+    AllocationMetadata * meta = addrToMetadata(slot_start);
```

Comment:
Can't it use simply `metadata[slot_index]` ?

Reply:
Yes, it should be fine here.